src/libpcp_web/src/query.c

```diff
@@ -3290,6 +3304,170 @@ series_calculate_standard_deviation(node_t *np)
+	np->value_set.series_values[i].series_desc.pmid = sdsnew(np->left->value_set.series_values[i].series_desc.pmid);
+	np->value_set.series_values[i].series_desc.semantics = sdsnew(np->left->value_set.series_values[i].series_desc.semantics);
+	np->value_set.series_values[i].series_desc.source = sdsnew(np->left->value_set.series_values[i].series_desc.source);
+	np->value_set.series_values[i].series_desc.type = sdsnew(np->left->value_set.series_values[i].series_desc.type);
+	np->value_set.series_values[i].series_desc.units = sdsnew(np->left->value_set.series_values[i].series_desc.units);
+    }
+}
+/*
+ * calculate the nth percentile series per-instance over time samples
+ */
+static void
+series_calculate_nth_percentile(node_t *np)
+{
+    seriesQueryBaton	*baton = (seriesQueryBaton *)np->baton;
+    unsigned int	n_series, n_samples, n_instances, i, j, k, l, m;
+    int			n, instance_idx, rank, *n_pointer;
+    double              *n_data, data, rank_d;
+    sds			msg;
+    pmSeriesValue       inst;
+
+    sscanf(np->right->value, "%d", &n);
+
+    n_series = np->left->value_set.num_series;
+    np->value_set.num_series = n_series;
+    np->value_set.series_values = (series_sample_set_t *)calloc(n_series, sizeof(series_sample_set_t));
+    for (i = 0; i < n_series; i++) {
+	n_samples = np->left->value_set.series_values[i].num_samples;
+	if (n_samples > 0) {
+	    np->value_set.series_values[i].num_samples = 1;
+	    np->value_set.series_values[i].series_sample = (series_instance_set_t *)calloc(1, sizeof(series_instance_set_t));
+	    n_instances = np->left->value_set.series_values[i].series_sample[0].num_instances;
+	    np->value_set.series_values[i].series_sample[0].num_instances = n_instances;
+	    np->value_set.series_values[i].series_sample[0].series_instance = (pmSeriesValue *)calloc(n_instances, sizeof(pmSeriesValue));
+	    rank_d = ((double)n/100 * n_samples);
+	    rank = (int) rank_d;
+	    for (k = 0; k < n_instances; k++) {
+                n_data = (double*) calloc(n_samples, sizeof(double));
+                n_pointer = (int*) calloc(n_samples, sizeof(int)); 
+
+		for (j = 1; j < n_samples; j++) {
+		    if (np->left->value_set.series_values[i].series_sample[j].num_instances != n_instances) {
+			if (pmDebugOptions.query && pmDebugOptions.desperate) {
+			    infofmt(msg, "number of instances in each sample are not equal\n");
+			    batoninfo(baton, PMLOG_ERROR, msg);
+			}
+			continue;
+		    }
+		    data = atof(np->left->value_set.series_values[i].series_sample[j].series_instance[k].data);
+		    for (l = 0; l < n_samples; ++l){
+			if (data > n_data[l]){
+			    for (m = n_samples - 1; m > l; --m){
+				n_data[m] = n_data[m-1];
+				n_pointer[m] = n_pointer[m-1];
+			    }
+			n_data[l] = data;
+			n_pointer[l] = j;
+			break;
+			}
+		    }
+		}
+		if (rank == n_samples){
+		    instance_idx = n_pointer[0];
+		}
+                else{
+		    instance_idx = n_pointer[n_samples-1-rank];
```

Comment:
I observed this line crash with some large (possibly invalid) input numbers - if 'rank' is greater than 'n_samples' this code will attempt to access memory outside of the n_pointer array.

Reply:
rank should not be greater than the n_sample, since the formula to calculate rank is (int)(n/100 * n_samples), where n is the percentile value. For example if n_samples = 10, n = 90, then rank = 9, which is the largest value. So the range will be 0 to n_samples.